firmware/defmt-test/macros/src/lib.rs

```diff
@@ -15,6 +15,16 @@ pub fn tests(args: TokenStream, input: TokenStream) -> TokenStream {
     }
 }
 
+macro_rules! fn_state_signature_msg {
+    ($name:literal) => {
+        concat!(
+            "`#[",
+            $name,
+            "]` function must have signature `fn()` or `fn(state: &mut T)`"
+        )
+    };
+}
```

Comment:
Could this also be a function? If not, it is okay as well. I just think functions are usually easier to understand.

```suggestion
fn fn_state_signature_msg(name: &str) -> String {
    format!("`#[{name}]` function must have signature `fn()` or `fn(state: &mut T)`")
}
```

Reply:
not sure how Rust optimizes format!() in this case, because String is heap allocated while the macro is compile time (so constant).

Thats why I went with a macro.

For defmt-test to be usable in no_std, it is not possible to use String.

Reply:
> For defmt-test to be usable in no_std, it is not possible to use String.

True, makes sense